File: controllers/AuthController.go
package controllers

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/dosemedia/pretest-api/prisma/db"
	"github.com/golang-jwt/jwt"
	"github.com/labstack/echo/v4"
	goredislib "github.com/redis/go-redis/v9"
)

type AuthController struct {
	client *db.PrismaClient
	cache  *goredislib.Client
}

func NewAuthController(e *echo.Echo) *AuthController {
	redisAddress := os.Getenv("REDIS_ADDRESS")
	if redisAddress == "" {
		redisAddress = "localhost:6379"
	}
	redisPassword := os.Getenv("REDIS_PASSWORD")
	redisDB := 0

	controller := AuthController{
		client: db.NewClient(),
		cache: goredislib.NewClient(&goredislib.Options{
			Addr:     redisAddress,
			Password: redisPassword,
			DB:       redisDB,
		}),
	}
	return &controller
}

func FlushRedisPrefix(client *goredislib.Client, prefix string) error {
	keys, err := client.Keys(context.Background(), prefix+"*").Result()
	if err != nil {
		return err
	}
	if len(keys) > 0 {
		_, err := client.Del(context.Background(), keys...).Result()
		if err != nil {
			return err
		}
	}
	return nil
}

func authToken(controller *AuthController, token string, c echo.Context) error {
	if token == "" {
		return c.JSON(http.StatusOK, map[string]interface{}{
			"X-Hasura-Role": "public",
		})
	}

	// Parse the token
	parsed, tokenError := jwt.Parse(token, func(token *jwt.Token) (interface{}, error) {
		keyStr := os.Getenv("JWT_TOKEN_KEY")
		if keyStr == "" {
			return nil, fmt.Errorf("backend not configured - missing jwt token key")
		}
		key := []byte(keyStr)
		return key, nil
	})
	if tokenError != nil {
		return c.JSON(http.StatusBadRequest, map[string]interface{}{
			"message": tokenError.Error(),
		})
	}

	if !parsed.Valid {
		return c.JSON(http.StatusUnauthorized, map[string]interface{}{
			"message": "Invalid token.",
		})
	}

	claims := parsed.Claims.(jwt.MapClaims)
	userId := claims["user_id"].(string)

	if userId == "" {
		return c.JSON(http.StatusUnauthorized, map[string]interface{}{
			"message": "Invalid token.",
		})
	}

	cacheKey := userId + ":" + token

	// Check for cached value in redis
	cachedResponseStr, cachedResponseError := controller.cache.Get(context.Background(), cacheKey).Bytes()
	if cachedResponseError == nil {
		cachedResponse := map[string]interface{}{}
		cachedResponseUnmarshalError := json.Unmarshal(cachedResponseStr, &cachedResponse)
		if cachedResponseUnmarshalError == nil {
			return c.JSON(http.StatusOK, cachedResponse)
		}
	}

	// No cached response for token, check if user exists
	user, userErr := controller.client.Users.FindUnique(db.Users.ID.Equals(userId)).Exec(context.Background())
	if userErr != nil {
		return c.JSON(http.StatusUnauthorized, map[string]interface{}{
			"message": "User not found.",
		})
	}

	// Use password timestamp to invalidate tokens when password changes
	// Note, this depends on user's cached tokens getting removed on password changes
	passwordAt := claims["password_at"].(string)
	if user.PasswordAt.String() != passwordAt {
		return c.JSON(http.StatusUnauthorized, map[string]interface{}{
			"message": "Token expired.",
		})
	}

	role := "user"
	roles := []string{role}

	if strings.Contains(user.Email, "@orchard-insights.com") {
		role = "superuser"
		roles = append(roles, "superuser")
	}

	response := map[string]interface{}{
		"X-Hasura-Role":          role,
		"X-Hasura-User-Id":       userId,
		"X-Hasura-Allowed-Roles": roles,
	}

	// Cache response
	responseJson, _ := json.Marshal(response)
	redisErr := controller.cache.Set(context.Background(), cacheKey, responseJson, time.Duration(time.Minute*60)).Err()
	if redisErr != nil {
		log.Print(redisErr)
	}

	return c.JSON(http.StatusOK, response)
}

func (controller *AuthController) Run(e *echo.Echo) error {
	if err := controller.client.Prisma.Connect(); err != nil {
		return err
	}

	// Check cache connection
	pingCmd := controller.cache.Ping(context.Background())
	pingErr := pingCmd.Err()
	if pingErr != nil {
		return pingErr
	}

	e.POST("/hasura/auth", func(c echo.Context) error {
		token := c.Request().Header.Get("Authorization")

		if token == "" {
			body := map[string]interface{}{}
			bodyErr := json.NewDecoder(c.Request().Body).Decode(&body)
			if bodyErr == nil {
				token = body["token"].(string)
			}
		}

		if strings.HasPrefix(token, "Bearer ") {
			token = strings.Replace(token, "Bearer ", "", 1)
		}

		return authToken(controller, token, c)
	})

	e.GET("/hasura/auth", func(c echo.Context) error {
		token := c.Request().Header.Get("Authorization")

		if token == "" {
			token = c.QueryParam("token")
		}

		if strings.HasPrefix(token, "Bearer ") {
			token = strings.Replace(token, "Bearer ", "", 1)
		}

		return authToken(controller, token, c)
	})

	return nil
}

func (controller *AuthController) Shutdown() {
	log.Print("Shutting down auth controller...")
	if controller.client != nil {
		controller.client.Disconnect()
	}
}

File: src/controllers/AuthController.ts
import { Express, Request, Response } from 'express'
import Controller from './Controller'
import _ from 'lodash'
import cache from '../cache'
import jwt from 'jsonwebtoken'
import prisma from '../database'
import { TokenPayload, verifyToken } from '../auth'

const cacheTTL = 1800

class AuthController implements Controller {
  startup (app: Express) {
    app.all('/hasura/auth', async (req: Request, res: Response) => {
      // Be nice in the way we look for a token

      // Check auth header first
      let token: string = req.headers.authorization || ''

      // Then look in request body
      if (!token && _.has(req, 'body')) {
        token = req.body.token || ''
      }

      // Then look in querystring
      if (!token && _.has(req, 'query.token')) {
        token = req.query.token + ''
      }

      // Strip "Bearer " off the token for decoding below
      if (_.startsWith(token, 'Bearer')) {
        token = token.replace('Bearer ', '')
      }

      // No token - you're public
      if (!token) {
        return res.json({
          'x-hasura-role': 'public'
        })
      }

      try {
        // Decode and verify the token
        const decoded = verifyToken(token)
        const userId = decoded.user_id
        const cacheKey = userId+':'+token

        if (!userId) {
          // No user id in the token => not authenticated
          res.status(401).json({ error: 'unauthorized' })
          return
        }

        // Check for cached response for this token to save round trips to db
        const cached = await cache.get(cacheKey)
        if (cached) {
          if (cached === 'unauthorized') {
            return res.status(401).json({ error: 'unauthorized' })
          }
          return res.json(JSON.parse(cached))
        }

        // Check if user exists
        const user = await prisma.users.findUnique({
          where: {
            id: userId,
          },
        })
        if (!user) {
          throw new Error('User not found!')
        }

        // Use password timestamp to invalidate tokens when password changes
        // Note, this depends on user's cached tokens getting removed on password changes
        if (user.password_at.getTime() !== decoded.password_at) {
          throw new Error('Token expired!')
        }

        const role = 'user'

        // Give hasura the user's id and role
        const responseData = {
          'X-Hasura-Role': role,
          'X-Hasura-User-Id': userId,
          'X-Hasura-User-Email': user.email
        }
        await cache.set(cacheKey, JSON.stringify(responseData), cacheTTL)
        res.json(responseData)

      } catch (error) {
        console.error(error)
        res.status(401).json({ error: 'unauthorized' })
      }
    })
  }

  async shutdown () {
    // No shutdown actions for this controller
  }
}

const controller = new AuthController()

export default controller
